### ISSUE TITLE
Model Bringup : LTX-2.3 [Fast, Pro]

### ISSUE THREAD

**Issue by saiarthiraguram** [tenstorrent/tt-xla#4542]
### Sources 

Reference : https://huggingface.co/Lightricks/LTX-2.3

**Comment by saiarthiraguram** [comment]
## LTX-2.3 bringup — status update (2026-09-07)

Consolidated update on this tracker. Two blockers previously recorded here have changed
character, one component is now root-caused, and one recorded state turned out not to be true.

### Component status

| Component | Status |
|---|---|
| Video VAE decoder | **PASS** — PCC 0.99, real weights, single device, bf16 |
| Video VAE encoder | **PASS** — PCC 0.99 |
| Audio VAE decoder | **PASS** — PCC 0.99 |
| Audio VAE encoder | **PASS** — PCC 0.99 |
| Vocoder | **FAIL, root-caused** — gated on #6009 (below) |
| Transformer (48L, TP=8) | **Compiles; fails at runtime in CCL** (below) |
| Gemma text encoder | Not started |
| Connector / caption projection | Checkpoints located — no longer blocked on availability |
| Spatial / temporal upsampler | Checkpoints located — standalone files |
| Full E2E (text → video + audio) | Not started |

All four VAE components were re-verified against the submodule conversion, not carried over from
the pre-conversion tree.

### Vocoder — root cause confirmed, and it is a handoff

Full analysis posted on **#6009**. Summary: the mechanism is **bf16 destination accumulation in
the device `Conv1d` reduction**. Device conv error grows with reduction length (relL2 0.0057 at
R=96 → 0.0729 at R=4608) while the CPU control stays flat at 0.0017, because CPU's `F.conv1d`
accumulates bf16 operands in fp32 — a **26× gap** at the vocoder's real R=2304. Confirmed
causally on the full model: holding `optimization_level=0` fixed and flipping only
`fp32_dest_acc_en`, PCC moves **0.94151 → 0.20610**.

**The fix on that axis is necessary but not sufficient.** The correct target is 0.988, not 1.0
(bf16 storage alone costs 0.012 PCC on CPU). With fp32 accumulation on, the model reaches
**0.94151 — 0.047 short of that ceiling and 0.048 short of the gate** — and math fidelity cannot
help because hifi4 is already the MLIR default. So this stays open as a tt-metal / tt-mlir handoff
with an op-level reproducer, not as a tt-xla fix.

The residual is now localized, which sharpens the handoff. At R=2304 with fp32 accumulation on,
**matmul reaches CPU parity (relL2 0.00168 vs 0.00165) but the conv stays 2.1× above CPU
(0.00353)**. Compounded over ~202 convolutions and amplified by the BWE stages, that is the right
order of magnitude for the missing 0.047 — so the open question is not "find a second defect" but
"why does the flag fix matmul accumulation and not conv accumulation?"

Two corrections that matter beyond this component:

- **The #5332 (Kokoro-82M) sine-phase link is refuted and has been dropped.** TT `sin` matches
  fp64 to 0.999999 out to |arg|=128, snake reads 0.999999, and the STFT rewrite reads 0.99990 on
  device. Kokoro is still a plausible candidate for the *conv-accumulation* mechanism, but at its
  real conv widths — this defect is invisible below roughly R=1000.
- **`fp32_dest_acc_en` and `math_fidelity` are both silently dropped at
  `optimization_level >= 1`.** At opt1, `True` and `False` produce byte-identical device output on
  both conv and matmul, while at opt0 the same flag is worth 12.6× / 25.7× in relative error.
  (For conv, opt1 output is byte-identical to opt0's explicit `False` — opt1 *is* forced bf16
  accumulation.) Being filed separately; opt1 is the default for model tests, so **any** accuracy
  experiment setting either flag at opt1 silently measures nothing and reports a false negative.
  This investigation produced exactly that false negative and nearly recorded it.

### Transformer — tt-mlir #9014 is closed, the blocker moved to runtime

**#9014 was already fixed upstream before it was filed.** The 1435 `sdy.manual_computation` ops
came from `WrapUnderManualComputationPass` wrapping every top-level func, private composite
`.impl` callees included; tt-mlir **#8601** (`7d6b6a51`, 2026-07-07) added
`if (funcOp.isPrivate()) return;` to that walk. The current pin contains it, and the 48-layer
failure mode does not recur.

With that cleared, the full 48-layer graph **traces and compiles** — 30m54s, mesh `(1, 8)` over 8
devices, executable submitted. Execution then fatals:

```
TT_FATAL @ tt_metal/impl/tensor/tensor_apis.cpp:631: buffers.size() == 1
  Can't get a single buffer from host storage distributed over mesh shape MeshShape([1, 8])
  in ccl::run(DistributeTensorOp) under FlatbufferLoadedExecutableInstance::execute
```

surfacing at exit as `Error code: 13`. That is the same PJRT status code as #9014's title but a
**different cause** — a runtime fatal after a good compile, not an ingest rejection. Please do not
reopen #9014 on the strength of the shared code.

Note also that two earlier environment faults, not compiler bugs, were responsible for the
apparent 48L blockage: a root-owned `.tt-metal-cache` directory, and a venv drifted to torch 2.10
against a pinned 2.11 (`add_id_match_guard()` arity, which only bites the multichip sharded
paths). Both are cleared in the current tree.

### `test_transformer_sharded` reports a false pass 
The test body ends with `with torch.no_grad(): compiled(*inputs)` — no `torch_xla.sync()`, no
`.cpu()`, no assertion. Under lazy XLA that only enqueues, so the body returns before the async
execution fatal lands: the log reads `1 passed` directly above the `TT_FATAL`. **As written it
would go green in CI on a model that never executed.** This also weakens the earlier "TP=8 proven
at 2L/24L" evidence, which came from the same non-materializing body. Being fixed, and the other
multichip sharded transformer tests (pattern copied from the Mochi TP MVP) need the same audit.

### Checkpoints 

Established by reading the HF listings plus the safetensors header directly:

- `ltx-2.3-22b-dev.safetensors` is **one flat 46 GB checkpoint of 5947 tensors**, not just the
  transformer: `model.*` 4444, `vocoder.*` 1227, `vae.*` 170, `audio_vae.*` 102,
  `text_embedding_projection.*` 4.
- **Caption projection** and **connector** are *inside* that checkpoint
  (`text_embedding_projection.*` ~1.15B params; `model.diffusion_model.*_embeddings_connector.*`,
  258 tensors). Upstream has no separate connector download because there is nothing separate to
  download. The `connectors/` artifact in the `Lightricks/LTX-2` HF repo is the diffusers-layout
  LTX-2 one, not the native path.
- **Spatial upscaler** is a standalone `ltx-2.3-spatial-upscaler-x2-1.1.safetensors` (1.00 GB) in
  the LTX-2.3 repo, plus x1.5 and a temporal x2. It never was inside the 22b checkpoint — which is
  why "not in the cached 22b-dev safetensors" was literally true and still sent people looking in
  the wrong place.
- **Gemma text encoder** is `google/gemma-3-12b-it-qat-q4_0-unquantized`, named by upstream. It is
  **not** the cached `Lightricks/LTX-2/text_encoder` — that is LTX-2's encoder and using it here
  would be wrong.

So the connector and upsampler work is unblocked; it was never a checkpoint-availability problem.

### Submodule conversion — landed on the branch

`ltx2_3/pytorch/src/` (53 files / 9069 lines of vendored `ltx_core`) is deleted. `loader.py`
registers `ltx_core` directly off `third_party/LTX-2` @ `7809842` with **no `sys.path` mutation**,
provides a `torchaudio` stub, and applies **4 drift-guarded runtime patches** so an upstream
change fails loudly instead of silently forking. Conventions follow tt-forge-models #776
(VibeVoice), which is on `main`.

### Corrections to previously recorded state

- **The vocoder xfail does not exist.** It was recorded as done ("`test_ltx2_3_vocoder` is xfailed
  against #6009 in #5714"), but `tests/torch/models/ltx2_3/test_audio_vae.py` on #5714 carries no
  marker. Verified 2026-09-07. The test fails outright today. Being added.
- **No LTX-2.3 entries exist in `tests/runner/test_config/`.** Nothing from this bringup is in any
  CI matrix, so every result here is a local run that will rot at the next uplift.

### PR chain

| PR | State |
|---|---|
| tt-forge-models **#796** — submodule conversion + loader | Ready for review |
| tt-xla **#5714** — component tests | Draft; strictly blocked on #796 merging + a pin bump |

### Next

1. Add the vocoder `KNOWN_FAILURE_XFAIL` against #6009 and register the four passing VAE
   components in the test-config YAML.
2. Fix `test_transformer_sharded` to materialize and assert, then re-measure the TP=8 result.
3. Get a tt-metal / tt-mlir owner for #6009 (conv accumulation precision + the ~0.047 residual).
4. File the `optimization_level >= 1` flag-drop issue.
5. Land tt-forge-models PR → bump the pin → take #5714 out of draft.
6. Then: Gemma text encoder, connector, spatial upsampler, full E2E.
